pyro/nn/auto_reg_nn.py

```diff
@@ -49,78 +119,67 @@ class AutoRegressiveNN(nn.Module):
-        else:
-            # the dependency structure is given by the user
-            self.mask_encoding = mask_encoding
+        self.num_layers = num_layers
 
         if permutation is None:
-            # a permutation is chosen at random
-            self.permutation = torch.randperm(input_dim, device=torch.device('cpu'))
+            # By default set a random permutation of variables, which is important for performance with multiple steps
+            self.permutation = torch.randperm(input_dim)
         else:
-            # the permutation is chosen by the user
-            self.permutation = permutation
-
-        # these masks control the autoregressive structure
-        self.mask1 = torch.zeros(hidden_dim, input_dim)
-        self.mask2 = torch.zeros(input_dim * self.output_dim_multiplier, hidden_dim)
-
-        for k in range(hidden_dim):
-            if self.mask_encoding.dim():
-                # fill in mask1
-                m_k = self.mask_encoding[k].item()
-            else:
-                m_k = self.mask_encoding.item()
-            slice_k = torch.cat([torch.ones(m_k), torch.zeros(input_dim - m_k)])
-            for j in range(input_dim):
-                self.mask1[k, self.permutation[j]] = slice_k[j]
-            # fill in mask2
-            slice_k = torch.cat([torch.zeros(m_k), torch.ones(input_dim - m_k)])
-            for r in range(self.output_dim_multiplier):
-                for j in range(input_dim):
-                    self.mask2[r * input_dim + self.permutation[j], k] = slice_k[j]
-
-        self.lin1 = MaskedLinear(input_dim, hidden_dim, self.mask1)
-        self.lin2 = MaskedLinear(hidden_dim, input_dim * output_dim_multiplier, self.mask2)
-        self.relu = nn.ReLU()
-
-    def get_mask_encoding(self):
-        """
-        Get the mask encoding associated with the neural network: basically the quantity m(k) in the MADE paper.
-        """
-        return self.mask_encoding
+            # The permutation is chosen by the user
+            self.permutation = permutation.type(dtype=torch.int64)
+
+        # Create masks
+        self.M_W, self.M_V, self.M_A = create_mask(input_dim=input_dim, observed_dim=0, hidden_dim=hidden_dim,
+                                                   num_layers=num_layers, permutation=self.permutation,
+                                                   output_dim_multiplier=output_dim_multiplier)
+
+        # Create masked layers
+        layers = [MaskedLinear(input_dim, hidden_dim, self.M_W[0])]
+        for i in range(1, num_layers):
+            layers.append(MaskedLinear(hidden_dim, hidden_dim, self.M_W[i]))
+        self.layers = nn.ModuleList(layers)
+
+        if skip_connections:
+            self.skip_p = MaskedLinear(input_dim, input_dim * output_dim_multiplier, self.M_A, bias=False)
+        else:
+            self.M_A = None
+            self.skip_p = None
+        self.p = MaskedLinear(hidden_dim, input_dim * output_dim_multiplier, self.M_V)
```

Comment:
can `self.p` be given a more transparent name?

Reply:
Yes and I think some of the other variables like the mask names can have more transparent names too

Reply:
These are all renamed now, and I made it a bit more concise by eliminating self.p